# Solving linear equations

$\newcommand{\A}{\mathbf{A}}$
$\newcommand{\B}{\mathbf{B}}$
$\newcommand{\C}{\mathbf{C}}$
$\newcommand{\D}{\mathbf{D}}$
$\newcommand{\x}{\mathbf{x}}$
$\newcommand{\v}{\mathbf{v}}$
$\newcommand{\L}{\mathbf{L}}$
$\newcommand{\U}{\mathbf{U}}$
$\newcommand{\V}{\mathbf{V}}$
$\newcommand{\Q}{\mathbf{Q}}$
$\newcommand{\R}{\mathbf{R}}$
$\newcommand{\Lzero}{\mathbf{L_0}}$
$\newcommand{\Lone}{\mathbf{L_1}}$
$\newcommand{\Ltwo}{\mathbf{L_2}}$

In physics, we often encounter systems of linear equations, such as
\begin{align}
    2x+3y &= 8\\
    8x-2y &= 4.
\end{align}

## Gaussian elimination

One approach to solving this system of equations is to multiply both sides of the first equation by -4, to obtain $-8x-12y = -32$.  If we add this to the second equation, the factors involving $x$ conveniently cancel one another and we are left with an equation depending only on $y$ ($-14y = -28$), which can be solved trivially to yield $y=2$.  Finally, we can plug this relation into either of the original two equations to solve and determine that $x=1$. 

To solve such systems of equations numerically, we can use a similar technique known as "Gaussian elimination."  Let's investigate this in the context of a system of three equations with three unknowns:
\begin{align}
    2x-2y+2z &= 16\\
    2x+3y-z &= -2\\
    3x-2y-9z &= 9.
\end{align}
  This system of equations can be represented in matrix form as
\begin{equation}
    \begin{pmatrix}
        2 & -2 & 2\\
        2 &  3 & -1\\
        3 &  -2 & -9
    \end{pmatrix}\begin{pmatrix}
        x\\y\\z
    \end{pmatrix} = \begin{pmatrix}
        16\\-2\\9
    \end{pmatrix}.
\end{equation}
Or, more simply, as $\A \x=\v$, where
\begin{equation}
    \A=\begin{pmatrix}
        2 & -2 & 2\\
        2 &  3 & -1\\
        3 &  -2 & -9
    \end{pmatrix},~ 
    \x=\begin{pmatrix}
        x \\ y \\ z
    \end{pmatrix},~\mathrm{and}~
    \v=\begin{pmatrix}
    16\\-2\\9
    \end{pmatrix}.
\end{equation}
To aide the discussion, let's label the entries of $\A$ by their row and column indices:
\begin{equation}
    \A=\begin{pmatrix}
            a_{0,0} & a_{0,1} & a_{0,2}\\
            a_{1,0} & a_{1,1} & a_{1,2}\\
            a_{2,0} & a_{2,1} & a_{2,2}\\
    \end{pmatrix}
\end{equation}

------------------------------
    
Gaussian elimination involves two repeated operations:
 1. Multiplication of the rows of the matrices $\A$ and $\v$ by a constant
 2. Subtraction of one row of these matrices from another
 
Using this procedure, just as above, we eventually arrive at an equation involving only a single variable which is trivial to solve.  By "backward substituting" this value, we can obtain the value of other variables.
        
Gaussian elimination begins by dividing the first row of matrices $\A$ and $\v$ by $a_{0,0}$, yielding: 
\begin{equation}
    \begin{pmatrix}
        1 & -1 & 1\\
        2 &  3 & -1\\
        3 &  -2 & -9
    \end{pmatrix}\begin{pmatrix}
        x\\y\\z
    \end{pmatrix} = \begin{pmatrix}
        8\\-2\\9
    \end{pmatrix}.
\end{equation}

Next, for each of the lower rows $i$, we subtract $a_{i,0}$ times the first row, so that the first element in each of these rows becomes zero.  In this case, we subtract 2 (3) times the first row from the second (third) row, yielding:
\begin{equation}
    \begin{pmatrix}
        1 & -1  & 1\\
        0 &  5 & -3\\
        0 &  1 & -12
    \end{pmatrix}\begin{pmatrix}
        x\\y\\z
    \end{pmatrix} = \begin{pmatrix}
        8\\-18\\-15
    \end{pmatrix}.
\end{equation}

Next, we move down one row and right one column, and repeat this process.  We divide the second row of matrices $\A$ and $\v$ by $a_{1,1}$, yielding: 
\begin{equation}
    \begin{pmatrix}
        1 & -1  & 1\\
        0 &  1 & -3/5\\
        0 &  1 & -12
    \end{pmatrix}\begin{pmatrix}
        x\\y\\z
    \end{pmatrix} = \begin{pmatrix}
        8\\-18/5\\-15
    \end{pmatrix}.
\end{equation}

For each of the lower rows $i$ (in this case, just the bottom row with $i=2$), we subtract $a_{i,1}$ times the second row, so that the second element in each of these rows becomes zero.  
\begin{equation}
    \begin{pmatrix}
        1 & -1  & 1\\
        0 &  1 & -3/5\\
        0 &  0 & -57/5
    \end{pmatrix}\begin{pmatrix}
        x\\y\\z
    \end{pmatrix} = \begin{pmatrix}
        8\\-18/5\\-57/5
    \end{pmatrix}.
\end{equation}
For the 3x3 matrix here, we are done.  For larger arrays, we continue with this process until we have an array with entries only in the upper triangle.

The third row of this matrix equation tells us that $\frac{-57}{5}z=\frac{-57}{5}$, or $z=1$.  

The second row tells us that $y-\frac{3}{5}z=-\frac{18}{5}$.  After backward substituting $z=1$, this becomes $y-\frac{3}{5}=-\frac{18}{5}$, or $y=-3$.

The first row tells us that $x-y+z=8$.  After backward substituting $y=-3$ and $z=1$, this becomes $x-(-3)+1=8$, or $x=4$.

Thus, we have fully solved the system of equations.

-------------------------------------------------------

Now, let's implment this algorithm using numpy.

In [1]:
import numpy as np

A = np.array([[ 2,  -2,  2 ],
              [ 2,   3, -1 ],
              [ 3,  -2, -9 ]], float)
v = np.array( [ 16, -2,  9 ] , float)

###########################################################################################

def backwardSub(A, v):
    N = len(v)
    
    x = np.zeros(N, float)

    for m in range(N-1, -1, -1):

        # Get the v value
        x[m] = v[m]/A[m,m]

        # First time through this loop is not entered.
        # In following m loops, we subtract each column entry of that row from the v value
        #  starting with the element to the right of the diagonal element
        for i in range(m+1,N):
            x[m] -= A[m,i]*x[i]/A[m,m]

    print("-------------")
    print()
    print("RESULT:")
    print("x =",x)
    print()
    
    return x
    
###########################################################################################

def solve(A, v):
    # N equations and N unknowns
    N = len(v)

    print("ORIGINAL ARRAYS:")
    print()
    print("A:")
    print(A)
    print("v:")
    print(v)
    print()

    for m in range(N):

        # Divide row m (of both A and v) by the diagonal entry A[m,m]
        div = A[m,m]
        A[m,:] /= div
        v[m] /= div

        print("DIVIDE ROW {0} BY A[{0},{0}] = {1}:".format(m,div))
        print()
        print("A:")
        print(A)
        print("v:")
        print(v)
        print()

        # For each subsequent row, subtract a multiple of row m such that mth column becomes zero
        for i in range(m+1,N):

            # Get the multiplier (the value of row i column m)
            mult = A[i,m]

            # Subtract mult times row m from row i (apply also to v)
            A[i,:] -= mult*A[m,:]
            v[i]   -= mult*v[m]

            print("SUBTRACT A[{2},{1}] = {0} TIMES ROW {1} FROM ROW {2}:".format(mult,m,i))
            print()
            print("A:")
            print(A)
            print("v:")
            print(v)
            print()

    ###########################################################################################
    
    backwardSub(A, v)

solve(A, v)

ORIGINAL ARRAYS:

A:
[[ 2. -2.  2.]
 [ 2.  3. -1.]
 [ 3. -2. -9.]]
v:
[16. -2.  9.]

DIVIDE ROW 0 BY A[0,0] = 2.0:

A:
[[ 1. -1.  1.]
 [ 2.  3. -1.]
 [ 3. -2. -9.]]
v:
[ 8. -2.  9.]

SUBTRACT A[1,0] = 2.0 TIMES ROW 0 FROM ROW 1:

A:
[[ 1. -1.  1.]
 [ 0.  5. -3.]
 [ 3. -2. -9.]]
v:
[  8. -18.   9.]

SUBTRACT A[2,0] = 3.0 TIMES ROW 0 FROM ROW 2:

A:
[[  1.  -1.   1.]
 [  0.   5.  -3.]
 [  0.   1. -12.]]
v:
[  8. -18. -15.]

DIVIDE ROW 1 BY A[1,1] = 5.0:

A:
[[  1.   -1.    1. ]
 [  0.    1.   -0.6]
 [  0.    1.  -12. ]]
v:
[  8.   -3.6 -15. ]

SUBTRACT A[2,1] = 1.0 TIMES ROW 1 FROM ROW 2:

A:
[[  1.   -1.    1. ]
 [  0.    1.   -0.6]
 [  0.    0.  -11.4]]
v:
[  8.   -3.6 -11.4]

DIVIDE ROW 2 BY A[2,2] = -11.4:

A:
[[ 1.  -1.   1. ]
 [ 0.   1.  -0.6]
 [-0.  -0.   1. ]]
v:
[ 8.  -3.6  1. ]

-------------

RESULT:
x = [ 4. -3.  1.]



Unfortunately, this simple alogrithm is not foolproof.  Let's apply the algorithm to the same array, but with $a_{1,0}=-2\rightarrow 3$.

In [2]:
A = np.array([[ 2,   3,  2 ],
              [ 2,   3, -1 ],
              [ 3,  -2, -9 ]], float)
v = np.array( [ 16, -2,  9 ],float)

solve(A, v)

ORIGINAL ARRAYS:

A:
[[ 2.  3.  2.]
 [ 2.  3. -1.]
 [ 3. -2. -9.]]
v:
[16. -2.  9.]

DIVIDE ROW 0 BY A[0,0] = 2.0:

A:
[[ 1.   1.5  1. ]
 [ 2.   3.  -1. ]
 [ 3.  -2.  -9. ]]
v:
[ 8. -2.  9.]

SUBTRACT A[1,0] = 2.0 TIMES ROW 0 FROM ROW 1:

A:
[[ 1.   1.5  1. ]
 [ 0.   0.  -3. ]
 [ 3.  -2.  -9. ]]
v:
[  8. -18.   9.]

SUBTRACT A[2,0] = 3.0 TIMES ROW 0 FROM ROW 2:

A:
[[  1.    1.5   1. ]
 [  0.    0.   -3. ]
 [  0.   -6.5 -12. ]]
v:
[  8. -18. -15.]

DIVIDE ROW 1 BY A[1,1] = 0.0:

A:
[[  1.    1.5   1. ]
 [  nan   nan  -inf]
 [  0.   -6.5 -12. ]]
v:
[  8. -inf -15.]

SUBTRACT A[2,1] = -6.5 TIMES ROW 1 FROM ROW 2:

A:
[[ 1.   1.5  1. ]
 [ nan  nan -inf]
 [ nan  nan -inf]]
v:
[  8. -inf -inf]

DIVIDE ROW 2 BY A[2,2] = -inf:

A:
[[ 1.   1.5  1. ]
 [ nan  nan -inf]
 [ nan  nan  nan]]
v:
[  8. -inf  nan]

-------------

RESULT:
x = [nan nan nan]



/var/folders/xj/h57tj5s562bf3y3sksq1yvsm0000gp/T/ipykernel_67134/2780258462.py:52: RuntimeWarning: divide by zero encountered in divide
  A[m,:] /= div
/var/folders/xj/h57tj5s562bf3y3sksq1yvsm0000gp/T/ipykernel_67134/2780258462.py:52: RuntimeWarning: invalid value encountered in divide
  A[m,:] /= div
/var/folders/xj/h57tj5s562bf3y3sksq1yvsm0000gp/T/ipykernel_67134/2780258462.py:53: RuntimeWarning: divide by zero encountered in scalar divide
  v[m] /= div
/var/folders/xj/h57tj5s562bf3y3sksq1yvsm0000gp/T/ipykernel_67134/2780258462.py:53: RuntimeWarning: invalid value encountered in scalar divide
  v[m] /= div


In this case, at an intermediate point of the algorithm, we encountered $a_{1,1}=0$ and attempted to divide row 1 by zero, with predictably disastrous consequences.

### Pivotting

The simplest approach to mitigate this issue is known as "pivoting."  At any point, if $a_{m,m}=0$, we simply swap row $m$ with a lower row.  Swapping rows can be achieved in numpy via:  

In [3]:
A = np.array([[ 1, 1, 1 ],
              [ 2, 2, 2 ],
              [ 3, 3, 3 ]], float)
m=1

A[[m,m+1],:] = A[[m+1,m],:]

A

array([[1., 1., 1.],
       [3., 3., 3.],
       [2., 2., 2.]])

In [4]:
import math

A = np.array([[ 2,   3,  2 ],
              [ 2,   3, -1 ],
              [ 3,  -2, -9 ]], float)
v = np.array( [ 16, -2,  9 ],  float)

def solveWPivot(A, v):
    # N equations and N unknowns
    N = len(v)

    print("ORIGINAL ARRAYS:")
    print()
    print("A:")
    print(A)
    print("v:")
    print(v)
    print()

    for m in range(N):

        # Divide row m (of both A and v) by the diagonal entry A[m,m]
        div = A[m,m]
        
        ################################################################
        #Pivot
        if math.isclose(div,0):
            v[[m,m+1]] = v[[m+1,m]]
            A[[m,m+1],:] = A[[m+1,m],:]
            div = A[m,m]
            print("*****SWAP ROWS {} AND {}*****".format(m,m+1))
            print()
            print("A:")
            print(A)
            print("v:")
            print(v)
            print()
        ################################################################

            
        A[m,:] /= div
        v[m] /= div

        print("DIVIDE ROW {0} BY A[{0},{0}] = {1}:".format(m,div))
        print()
        print("A:")
        print(A)
        print("v:")
        print(v)
        print()

        # For each subsequent row, subtract a multiple of row m such that mth column becomes zero
        for i in range(m+1,N):

            # Get the multiplier (the value of row i column m)
            mult = A[i,m]

            # For each column in row i, subtract mult times the value of that column entry in row m (apply also to v)
            A[i,:] -= mult*A[m,:]
            v[i]   -= mult*v[m]

            print("SUBTRACT A[{2},{1}] = {0} TIMES ROW {1} FROM ROW {2}:".format(mult,m,i))
            print()
            print("A:")
            print(A)
            print("v:")
            print(v)
            print()

    ###########################################################################################
    
    backwardSub(A, v)

solveWPivot(A, v)

ORIGINAL ARRAYS:

A:
[[ 2.  3.  2.]
 [ 2.  3. -1.]
 [ 3. -2. -9.]]
v:
[16. -2.  9.]

DIVIDE ROW 0 BY A[0,0] = 2.0:

A:
[[ 1.   1.5  1. ]
 [ 2.   3.  -1. ]
 [ 3.  -2.  -9. ]]
v:
[ 8. -2.  9.]

SUBTRACT A[1,0] = 2.0 TIMES ROW 0 FROM ROW 1:

A:
[[ 1.   1.5  1. ]
 [ 0.   0.  -3. ]
 [ 3.  -2.  -9. ]]
v:
[  8. -18.   9.]

SUBTRACT A[2,0] = 3.0 TIMES ROW 0 FROM ROW 2:

A:
[[  1.    1.5   1. ]
 [  0.    0.   -3. ]
 [  0.   -6.5 -12. ]]
v:
[  8. -18. -15.]

*****SWAP ROWS 1 AND 2*****

A:
[[  1.    1.5   1. ]
 [  0.   -6.5 -12. ]
 [  0.    0.   -3. ]]
v:
[  8. -15. -18.]

DIVIDE ROW 1 BY A[1,1] = -6.5:

A:
[[ 1.          1.5         1.        ]
 [-0.          1.          1.84615385]
 [ 0.          0.         -3.        ]]
v:
[  8.           2.30769231 -18.        ]

SUBTRACT A[2,1] = 0.0 TIMES ROW 1 FROM ROW 2:

A:
[[ 1.          1.5         1.        ]
 [-0.          1.          1.84615385]
 [ 0.          0.         -3.        ]]
v:
[  8.           2.30769231 -18.        ]

DIVIDE ROW 2 BY A[2

As you can easily confirm via substitution, these are the correct solutions.

However, pivoting has to be done carefully, as it can sometimes introduce new issues.  Therefore, the standard technique is "partial pivotting."  In this approach, we consider swapping rows even if $a_{m,m}\ne0$.  If the $m$th element of row $m$ is less than the $m$th element *in any lower row*, we swap row $m$ with the row containing the largest such value.

In [5]:
A = np.array([[ 2,   3,  2 ],
              [ 2,   3, -1 ],
              [ 3,  -2, -9 ]], float)
v = np.array( [ 16, -2,  9 ],  float)

m = 0

#print mth column (from row m to bottom) and max value in this column
print(    A[m:,m], 
      max(A[m:,m]))

#swap rows
np.where(A[m:,m] == max(A[m:,m]))[0][0]

[2. 2. 3.] 3.0


np.int64(2)

In [6]:
A = np.array([[ 2,   3,  2 ],
              [ 2,   3, -1 ],
              [ 3,  -2, -9 ]], float)
v = np.array( [ 16, -2,  9 ],  float)

def solveWPivot(A, v):
    # N equations and N unknowns
    N = len(v)

    print("ORIGINAL ARRAYS:")
    print()
    print("A:")
    print(A)
    print("v:")
    print(v)
    print()

    for m in range(N):

        ################################################################
        #Pivot
        swap = np.where(A[m:,m] == max(A[m:,m]))[0][0]
        #print(swap)
        if swap != 0:

            v[[m,swap]] = v[[swap,m]]
            A[[m,swap],:] = A[[swap,m],:]
            ############################################################
            div = A[m,m]
            print("*****SWAP ROWS {} AND {}*****".format(m,swap))
            print()
            print("A:")
            print(A)
            print("v:")
            print(v)
            print()
        
        # Divide row m (of both A and v) by the diagonal entry A[m,m]
        div = A[m,m]
            
        A[m,:] /= div
        v[m] /= div

        print("DIVIDE ROW {0} BY A[{0},{0}] = {1}:".format(m,div))
        print()
        print("A:")
        print(A)
        print("v:")
        print(v)
        print()

        # For each subsequent row, subtract a multiple of row m such that mth column becomes zero
        for i in range(m+1,N):

            # Get the multiplier (the value of row i column m)
            mult = A[i,m]

            # For each column in row i, subtract mult times the value of that column entry in row m (apply also to v)
            A[i,:] -= mult*A[m,:]
            v[i]   -= mult*v[m]

            print("SUBTRACT A[{2},{1}] = {0} TIMES ROW {1} FROM ROW {2}:".format(mult,m,i))
            print()
            print("A:")
            print(A)
            print("v:")
            print(v)
            print()

    ###########################################################################################
    
    backwardSub(A, v)

solveWPivot(A, v)

ORIGINAL ARRAYS:

A:
[[ 2.  3.  2.]
 [ 2.  3. -1.]
 [ 3. -2. -9.]]
v:
[16. -2.  9.]

*****SWAP ROWS 0 AND 2*****

A:
[[ 3. -2. -9.]
 [ 2.  3. -1.]
 [ 2.  3.  2.]]
v:
[ 9. -2. 16.]

DIVIDE ROW 0 BY A[0,0] = 3.0:

A:
[[ 1.         -0.66666667 -3.        ]
 [ 2.          3.         -1.        ]
 [ 2.          3.          2.        ]]
v:
[ 3. -2. 16.]

SUBTRACT A[1,0] = 2.0 TIMES ROW 0 FROM ROW 1:

A:
[[ 1.         -0.66666667 -3.        ]
 [ 0.          4.33333333  5.        ]
 [ 2.          3.          2.        ]]
v:
[ 3. -8. 16.]

SUBTRACT A[2,0] = 2.0 TIMES ROW 0 FROM ROW 2:

A:
[[ 1.         -0.66666667 -3.        ]
 [ 0.          4.33333333  5.        ]
 [ 0.          4.33333333  8.        ]]
v:
[ 3. -8. 10.]

DIVIDE ROW 1 BY A[1,1] = 4.333333333333333:

A:
[[ 1.         -0.66666667 -3.        ]
 [ 0.          1.          1.15384615]
 [ 0.          4.33333333  8.        ]]
v:
[ 3.         -1.84615385 10.        ]

SUBTRACT A[2,1] = 4.333333333333333 TIMES ROW 1 FROM ROW 2:

A:
[[ 1.

As they should, these are the same values that we obtained with standard pivoting.  However, this algorithm (with partial pivoting) is more robust.

--------------

After the first step of Gaussian elimination, we obtain a new matrix $B$ where $b_{0,0}=1$ and $b_{i,0} = 0$ for $i>0$.  It turns out that this transformation can be accomplished by multiplying $\A$ by the lower diagonal matrix
\begin{equation}
    \Lzero = \frac{1}{a_{0,0}}\times\begin{pmatrix}
        1 & 0  & 0\\
        -a_{1,0} &  a_{0,0} & 0\\
        -a_{2,0} &  0 & a_{0,0}\\
    \end{pmatrix}.
\end{equation}

This is easily verified using numpy.

In [7]:
A = np.array([[ 2,  -2,  2 ],
              [ 2,   3, -1 ],
              [ 3,  -2, -9 ]], float)
v = np.array( [ 16, -2,  9 ] , float)

L0 = (1/A[0,0]) * np.array([[ 1,      0,      0      ],
                            [-A[1,0], A[0,0], 0      ], 
                            [-A[2,0], 0,      A[0,0] ]], float)

B = np.matmul(L0,A)

B

array([[  1.,  -1.,   1.],
       [  0.,   5.,  -3.],
       [  0.,   1., -12.]])

Each of the subsequent steps of Gaussian elimination can similarly be represented by an appropriately chosen lower diagonal matrix.  The second step can be achieved by multipliying matrix $\B$ by
\begin{equation}
    \Lone = \frac{1}{b_{1,1}}\times\begin{pmatrix}
        b_{1,1} & 0        & 0\\
        0       & 1        & 0\\
        0       & -b_{2,1} & b_{1,1}\\
    \end{pmatrix}, 
\end{equation}resulting in matrix $\C$.

In [8]:
L1 = (1/B[1,1]) * np.array([[ B[1,1], 0,       0      ],
                            [ 0,      1,       0      ], 
                            [ 0,      -B[2,1], B[1,1] ]], float)

C = np.matmul(L1,B)

C

array([[  1. ,  -1. ,   1. ],
       [  0. ,   1. ,  -0.6],
       [  0. ,   0. , -11.4]])

The third step can be achieved by multipliying matrix $\C$ by
\begin{equation}
    \Ltwo = \frac{1}{c_{2,2}}\times\begin{pmatrix}
        c_{2,2} & 0        & 0\\
        0       & c_{2,2}  & 0\\
        0       & 0        & 1\\
    \end{pmatrix}, 
\end{equation}resulting in matrix $\D$.

In [9]:
L2 = (1/C[2,2]) * np.array([[ C[2,2], 0,       0      ],
                            [ 0,      C[2,2],  0      ], 
                            [ 0,      0,       1      ]], float)

D = np.matmul(L2,C)

D

array([[ 1. , -1. ,  1. ],
       [ 0. ,  1. , -0.6],
       [ 0. ,  0. ,  1. ]])

Successive multiplication of $\A$ by $\Ltwo\Lone\Lzero$ resulted in the same upper-diagonal matrix that we obtained above via Gaussian elimination.  

Let's apply the same transformation to vector $\v$:

In [10]:
np.matmul(L2,np.matmul(L1,np.matmul(L0,v)))

array([ 8. , -3.6,  1. ])

This is also the same vector that we obtained above through the process of Gaussian elimination.  

So putting this all together, we see that the process of Gaussian elimination is equivalent to multiplying both sides of the matrix equation we wish to solve, $\A \x=\v$, by $\Ltwo\Lone\Lzero$.

## Lower upper (LU) decomposition

We can take this matrix-based formulation of Gaussian elimination a step further.  

Let's define $\L=\Lzero^{-1}\Lone^{-1}\Ltwo^{-1}$ and $\U=\Ltwo\Lone\Lzero\A$.  We can see that these matrices are lower- and upper-diagonal, respectively (hence the names $\L$ and $\U$).

In [11]:
A = np.array([[ 2,  -2,  2 ],
              [ 2,   3, -1 ],
              [ 3,  -2, -9 ]], float)
v = np.array( [ 16, -2,  9 ] , float)

L = np.matmul( np.matmul(np.linalg.inv(L0),np.linalg.inv(L1)), np.linalg.inv(L2))
U = np.matmul( np.matmul( np.matmul(L2,L1), L0), A)

L,U

(array([[  2. ,   0. ,   0. ],
        [  2. ,   5. ,   0. ],
        [  3. ,   1. , -11.4]]),
 array([[ 1. , -1. ,  1. ],
        [ 0. ,  1. , -0.6],
        [ 0. ,  0. ,  1. ]]))

The product of these matrices is just our matrix $\A$.  Therefore, we can write the matrix equation we wish to solve as $\L\U \x=\v$.  In general, we can write this equation as
\begin{equation}\tag{1}
    \begin{pmatrix}
        l_{0,0} & 0       & 0\\
        l_{1,0} & l_{1,1} & 0\\
        l_{2,0} & l_{1,2} & l_{2,2}
    \end{pmatrix}
    \begin{pmatrix}
        u_{0,0} & u_{0,1} & u_{0,2}\\
        0       & u_{1,1} & u_{1,2}\\
        0       & 0       & u_{2,2}
    \end{pmatrix}
    \begin{pmatrix}
        x_0 \\
        x_1 \\
        x_2
    \end{pmatrix}=
    \begin{pmatrix}
        v_0 \\
        v_1 \\
        v_2
    \end{pmatrix}.
\end{equation}
Note the slight change of notation with respect to the preceeding section.  Here, the vector $\mathbf{x}\equiv(x_0, x_1, x_2) = (x, y, z)$.


Let's define a new vector $\mathbf{y}$, defined by
\begin{equation}\tag{2}
    \begin{pmatrix}
        u_{0,0} & u_{0,1} & u_{0,2}\\
        0       & u_{1,1} & u_{1,2}\\
        0       & 0       & u_{2,2}
    \end{pmatrix}
    \begin{pmatrix}
        x_0 \\
        x_1 \\
        x_2
    \end{pmatrix}=
    \begin{pmatrix}
        y_0 \\
        y_1 \\
        y_2
    \end{pmatrix}.
\end{equation}

In terms of this vector, we can write Equation 1 as
\begin{equation}\tag{3}
    \begin{pmatrix}
        l_{0,0} & 0       & 0\\
        l_{1,0} & l_{1,1} & 0\\
        l_{2,0} & l_{1,2} & l_{2,2}
    \end{pmatrix}
    \begin{pmatrix}
        y_0 \\
        y_1 \\
        y_2
    \end{pmatrix}=
    \begin{pmatrix}
        v_0 \\
        v_1 \\
        v_2
    \end{pmatrix}.
\end{equation}

Equation 2 involves an upper-diagonal matrix, which can be solved easily via backward substitution.  Equation 3 involves a lower-diagonal matrix, which can similarly be solved easily via forward substitution.  

Thus, the problem has been reduced to finding $\mathbf{y}$ from Equation 3 via forward substitution, then using these values to find $\x$ from Equation 2 via backward substitution.

In [12]:
def forwardSub(A, v):
    N = len(v)
    
    y = np.zeros(N, float)

    for m in range(0,N):

        # Get the v value
        y[m] = v[m]/A[m,m]

        # First time through this loop is not entered.
        # In following m loops, we subtract each column entry of that row from the v value
        #  starting with the element to the right of the diagonal element
        for i in range(0,m):
            y[m] -= A[m,i]*y[i]/A[m,m]

    print("-------------")
    print()
    print("RESULT:")
    print("y =",y)
    print()
    
    return y
    
###########################################################################################

y = forwardSub(L,v)
x = backwardSub(U,y)

-------------

RESULT:
y = [ 8.  -3.6  1. ]

-------------

RESULT:
x = [ 4. -3.  1.]



Using this method, we get the same (correct) solution that we obtained via Gaussian elimination above.  NB: this method is prone to issues and pivoting should generally be utilized.

But what is the advantage?  Recall that the calculation of $\L$ and $\U$ above depends only on the matrix $\A$, but not the particular vector $\v$.  It turns out that we often wish to solve many different sets of matrix equations with the same matrix $\A$ but different vectors $\v$.  With this method, we only need calculate the $\L$ and $\U$ matrices once, and use these matrices to solve for different vectors $\v_1, \v_2$, etc.

In [13]:
v = np.array( [ 1, 2, 3 ] , float)

y = forwardSub(L,v)
x = backwardSub(U,y)

#################################################################################

v = np.array( [ 9, 5,  1 ] , float)

y = forwardSub(L,v)
x = backwardSub(U,y)

-------------

RESULT:
y = [ 0.5         0.2        -0.11403509]

-------------

RESULT:
x = [ 0.74561404  0.13157895 -0.11403509]

-------------

RESULT:
y = [ 4.5        -0.8         1.02631579]

-------------

RESULT:
x = [ 3.28947368 -0.18421053  1.02631579]



These are of course the correct solutions for $\v = (1, 2, 3)$ and $\v = (9, 5, 1)$, respectively.

-----------

Now that we understand the basic procedure, we can also rely on numpy to do the heavy lifting for us.  There is a `solve()` function within the `linalg` submodule of numpy for exactly this type of problem.

In [14]:
A = np.array([[ 2,  -2,  2 ],
              [ 2,   3, -1 ],
              [ 3,  -2, -9 ]], float)
v = np.array( [ 9, 5,  1 ]   , float)

x = np.linalg.solve(A,v)

x

array([ 3.28947368, -0.18421053,  1.02631579])

## Eigenvalue problems

A similar class of problems that arise frequently within physics are known as "eigenvalue problems."  We are often given an $n\times n$ matrix $\A$ (usually symmetric or Hermitian) and asked to determine the set of $n$ eigenvectors $\v_1\dots\v_n$ and associated eigenvalues $\lambda_1\dots\lambda_n$ that satisfy the relation
\begin{equation}\tag{4}
    \A\v_i = \lambda_i\v_i.
\end{equation}
The eigenvectors are orthogonal to one another.

For an $n\times n$ matrix $\A$, Equation 4 really represents $n$ equations (one for each eigenvector/eigenvalue pair).  We can combine these three equations into a single matrix equation by defining matrix $\V$ with columns given by the eigenvectors, and diagonal matrix $\D$ with the eigenvalues as the diagonal elements.  
\begin{equation}\tag{5}
    \A\V = \V\D.
\end{equation}
Since the eigenvectors are orthogonal, matrix $\V$ is orthogonal, and $\V^T=\V^{-1}$.

-----------

A common method for solving such matrix equations relies on "QR decomposition."  Any real square matrix can be decomposed into the product of a orthogonal matrix $\Q_1$ and an upper-diagonal matrix $\R_1$: 
\begin{equation}
    \A=\Q_1\R_1.
\end{equation}
Multiplying this equation from the left by $\Q_1^T$ yields
\begin{equation}\tag{6}
    \Q_1^T\A=\Q_1^T\Q_1\R_1=\R_1,
\end{equation}
where we have made use of the fact that $\Q_1$ is orthogonal and therefore $\Q_1^T=\Q_1^{-1}$.

Now, let's define a new matrix $\A_1=\R_1\Q_1$.  Using Equation 6, we can replace $\R_1$ with $\Q_1^T\A$, yielding:
\begin{equation}
    \A_1=\Q_1^T\A\Q_1.
\end{equation}

We then repeat this process for $\A_1$ (which remains real and symmetric).  We perform QR decomposition of matrix $\A_1=\Q_2\R_2$ and define the matrix
\begin{align}
    \A_2&=\R_2\Q_2 \\
    &=\Q_2^T\A_1\Q_2 \\
    &=\Q_2^T\Q_1^T\A\Q_1\Q_2
\end{align}
If we continue this process, we eventually obtain matrix
\begin{equation}\tag{7}
    \A_k=(\Q_k^T\dots\Q_1^T)\A(\Q_1\dots\Q_k).
\end{equation}

With each iteration of this process, the matrix $\A_k$ becomes more diagonal, and we can eventually stop when the off-diagonal elements get sufficiently close to zero.

Let's define one additional matrix $\V=\prod_{i=1}^k\Q_i$.  Then, Equation 7 becomes $\V^T\A\V=\A_k$.  Multiplying this equation from the left by $\V$, we obtain $\A\V=\V\A_k$.  This is the same as Equation 5, with the replacement $\D\rightarrow\A_k$.  Thus, we can identify the matrix $\A_k$ as the diagonal matrix containing the desired eigenvalues.  Further, the columns of $\V$ represent the corresponding eigenvectors.

-------------

Let's explore this procedure using numpy, and find the eigenvalues and corresponding eigenvectors of the following matrix:

In [15]:
A = np.array([[ 2, 3, 5 ],
              [ 3, 4, 6 ],
              [ 5, 6, 7 ]], float)

A

array([[2., 3., 5.],
       [3., 4., 6.],
       [5., 6., 7.]])

We begin by decomposing the matrix $\A$ into the product of orthogonal matrix $\Q_1$ and upper-diagonal matrix $\R_1$.

In [16]:
Q1,R1 = np.linalg.qr(A)

Q1,R1

(array([[-0.32444284, -0.78039897, -0.53452248],
        [-0.48666426, -0.34684399,  0.80178373],
        [-0.81110711,  0.52026598, -0.26726124]]),
 array([[ -6.164414  ,  -7.78662821, -10.21994953],
        [  0.        ,  -0.60697698,  -2.34119692],
        [  0.        ,   0.        ,   0.26726124]]))

As a sanity check, let's confirm that the matrix $\Q_1$ is orthogonal.  If so, the product of different columns would be zero.

In [17]:
np.matmul(Q1[:,0],Q1[:,1]), np.matmul(Q1[:,0],Q1[:,0])

(np.float64(-9.296934019761335e-17), np.float64(1.0))

We see that this is indeed the case.

Next, we obtain the matrix $\A_1$ from the product $\R_1\Q_1$.

In [18]:
A1 = np.matmul(R1,Q1)

print("Maximum off diagonal element:",max(np.abs([A1[i,j] for i in range(3) for j in range(3) if i!=j])))

A1

Maximum off diagonal element: 2.1943554599125386


array([[14.07894737,  2.19435546, -0.21677749],
       [ 2.19435546, -1.0075188 ,  0.13904693],
       [-0.21677749,  0.13904693, -0.07142857]])

As claimed, the maximum off-diagonal element has been reduced from 6 to 2.2.

Next, we decompose the matrix $\A_1$ into $\Q_2\R_2$, and obtain the matrix $\A_2$ from the product $\R_2\Q_2$.

In [19]:
Q2,R2 = np.linalg.qr(A1)

A2 = np.matmul(R2,Q2)

print("Maximum off diagonal element:",max(np.abs([A2[i,j] for i in range(3) for j in range(3) if i!=j])))

A2

Maximum off diagonal element: 0.2097018626753977


array([[ 1.43914734e+01, -2.09701863e-01, -7.94128845e-04],
       [-2.09701863e-01, -1.33969218e+00, -6.58811382e-03],
       [-7.94128845e-04, -6.58811382e-03, -5.17811953e-02]])

The maximum off-diagonal element has now been reduced to just 0.2.

We could continue like this until the off-diagonal elements become sufficiently small.  However, it is more efficient to perform this procedure using a loop.

In [20]:
Ak = A
V = np.identity(3)

maxOffDiag = max(np.abs( [Ak[i,j] for i in range(3) for j in range(3) if i!=j] ))
print(Ak)
print()
print("Maximum off diagonal element:",maxOffDiag)
print()
print(50*"-")

while(maxOffDiag > 1e-10):
    Qk,Rk = np.linalg.qr(Ak)
    Ak = np.matmul(Rk,Qk)
    maxOffDiag = max(np.abs( [Ak[i,j] for i in range(3) for j in range(3) if i!=j] ))
    
    V = np.matmul(V,Qk)
    
    print()
    print(Ak)
    print()
    print("Maximum off diagonal element:",maxOffDiag)
    print()
    print(50*"-")
    
print(50*"-")
print(50*"-")
print()

print("Ak:")
print(Ak)
print()
print("V:")
print(V)

[[2. 3. 5.]
 [3. 4. 6.]
 [5. 6. 7.]]

Maximum off diagonal element: 6.0

--------------------------------------------------

[[14.07894737  2.19435546 -0.21677749]
 [ 2.19435546 -1.0075188   0.13904693]
 [-0.21677749  0.13904693 -0.07142857]]

Maximum off diagonal element: 2.1943554599125386

--------------------------------------------------

[[ 1.43914734e+01 -2.09701863e-01 -7.94128845e-04]
 [-2.09701863e-01 -1.33969218e+00 -6.58811382e-03]
 [-7.94128845e-04 -6.58811382e-03 -5.17811953e-02]]

Maximum off diagonal element: 0.2097018626753977

--------------------------------------------------

[[ 1.43942440e+01  1.95619848e-02 -2.85518767e-06]
 [ 1.95619848e-02 -1.34249644e+00  2.54327687e-04]
 [-2.85518767e-06  2.54327687e-04 -5.17475511e-02]]

Maximum off diagonal element: 0.019561984783908228

--------------------------------------------------

[[ 1.43942681e+01 -1.82450483e-03 -1.02644298e-08]
 [-1.82450483e-03 -1.34252059e+00 -9.80319612e-06]
 [-1.02644289e-08 -9.80319612e-06 -5

The off-diagonal elements are now less than the desired tolerance ($1\times10^{-10}$).  We see that the eigenvalues are 14.4, -1.34, and -0.0517.  These eigenvalues correspond to eigenvectors $v_1=(0.424, 0.541, 0.726)$, $v_2=(0.650, 0.377, -0.660)$, $v_3=(-0.631, 0.752, -0.192)$, respectively.

--------------

Of course, since this is a common class of problems, tools exist to perform this procedure for you.  Within numpy's `linalg` module, there is an `eigh()` function for calculating eigenvalues and eigenvectors of symmetric or Hermitian matrices.

In [21]:
x,V = np.linalg.eigh(A)

x,V

(array([-1.34252081, -0.0517475 , 14.39426831]),
 array([[-0.64954563,  0.63116053, -0.42394205],
        [-0.37687174, -0.75153297, -0.54144795],
        [ 0.660347  ,  0.19192337, -0.72602153]]))

As it should, the eigenvalues and eigenvectors found with numpy agree with those that we calculated above.

There is also an `eig()` function that should be used for matrixes which are not symmetric or Hermitian.

In [22]:
A = np.array([[ 2,  -2,  2 ],
              [ 2,   3, -1 ],
              [ 3,  -2, -9 ]], float)

x,V = np.linalg.eig(A)

x,V

(array([-9.66759246+0.j        ,  2.83379623+1.93947752j,
         2.83379623-1.93947752j]),
 array([[ 0.15115371+0.j        ,  0.73167617+0.j        ,
          0.73167617-0.j        ],
        [-0.10148686+0.j        , -0.09206753-0.63681248j,
         -0.09206753+0.63681248j],
        [-0.98328682+0.j        ,  0.21296689+0.07272226j,
          0.21296689-0.07272226j]]))